File: spikes/rust/src/deleted.rs
//! A row in the trash: the domain object it was, and when it went.
//!
//! One generic type where the OCaml spike writes `Issue.deleted` and
//! `Project.deleted` as two records. `Group<Deleted<Issue>>` and
//! `Group<Deleted<Project>>` are then distinct types for free, which is what
//! makes "a deleted row cannot be edited" a fact about what compiles rather
//! than a rule the registration lists happen to follow.

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Deleted<T> {
    pub inner: T,
    pub deleted_at: String,
}

File: spikes/python/src/tt/platform/deleted.py
"""A row in the trash: the domain object it was, and when it went.

One generic type, so ``Group[Deleted[Issue]]`` and ``Group[Deleted[Project]]``
are distinct types — which is what makes "a deleted row cannot be edited" a fact
about the registration's type rather than a rule the lists happen to follow.
"""

from dataclasses import dataclass


@dataclass(frozen=True)
class Deleted[T]:
    inner: T
    deleted_at: str
